### DISCUSSION TITLE
Slack workspace

### DISCUSSION THREAD

**Discussion by attipaci** [nom-tam-fits/nom-tam-fits#204] [Ideas]
I have created a Slack workspace where we can collaborate discuss more easily. If you are interested to join us on Slack, send me a request with the e-mail address and I'll send you an invite...

@ritchieGitHub @wcleveland @olebole @vforchi @Zlika @mbtaylor

**Comment by olebole** [comment]
I usually would prefer to keep things on Github, since then we have everything on one place, and it is safe for later reference. IMO the discussions we had so far (Bug reports, patches as those from Starjava) fit better into Github issues and pull requests than into Slack discussions.
But this is your decision; if you like to have slack, please invite me with olebole@debian.org. However, I don't have slack constantly open, so my response time may be a few days.

**Comment by attipaci** [comment]
I don't want to move discussion away from here. And anything that explains or clarifies problems or changes should stay on github (or be copied) here. I'm more thinking of Slack as an option for quick exchanges, questions, or brainstorming ideas. It's up to you if and how you want to use it...

**Comment by wcleveland** [comment]
As a slack user, I agree it can make quick exchanges easier. That said, it was the increased activity in Github that caught my attention and led me to check and see what is going on. 😉

**Comment by attipaci** [comment]
@wcleveland, I'm glad you are joining the conversation.Let me know if you want me to add you to Slack (with what email?).

I have been working on simplifying and making header parsing and creating more robust. I did like the work you did on scientific notation, but there were some nagging inconsistencies left (such as issue #169, and the somewhat related #165). I'm now working on a grounds-up overhaul for a lot of the HeaderCard functionality (same API, just cleaner implementation). When I'm done, I'd very much appreciate your feedback, since they will affect a fair bit of the changes you proposed in your PRs. Slack might be a good way to interract, but we can keep it strictly in the tickets also. Anyway, welcome back!

**Comment by wcleveland** [comment]
As long as you keep this work in a separate branch and allow time for verification (and sign off) before merging to main then I'm all for code cleanup.

**Comment by attipaci** [comment]
@wcleveland, When you are up for it, have a look at PR #170. The PR has a good summary of the changes (quite a lot!). It also points you to the branch on my fork that you can check out to do your own testing. Don't get scared by the long list of changes. They don't much affect API compared to the last release (1.15.2), and when they do (such as the new exceptions) it happens only when the FITS file would be definitely corrupted. Since probably nobody is using the library for creating corrupted FITS files, the new exceptions aren't at all likely to break good working code (indeed, they did not break ANY of the unit tests!). But, they will help debugging new applications, for sure, by preventing stupid mistakes.

The biggest changes are to the still unreleased API from handling of numbers and precision control (by you!). I have mainly reorganized and consolidated what you have done, simplified it, and fixed some inconsistencies. For example, 

- The used of 'D' (instead of 'E') is now controlled by settings in `FitsFactory` (similar to other switches), and 'D' is used only when the precision warrants it. Conversely `D` is recognised when parsing, and bumps the value type to `double` even if the number of digits would otherwise fit into a `float`.
- Also, 'precision' is now universally understood as the _number of decimal places after the leading figure_ (which is how it was used in the exponential notation). Previously the use of absolute precision in some cases (fixed format), and relative precision in other cases (exponential format) was __VERY__ confusing, and the switch between then could happen under the hood (such as in the default formatting flow) -- making it also __VERY__ unpredictable... 
- All number formatting is now handled by a `nom.tam.util.FlexFormat`.
- Methods use the generic `Number` instead of `int`, `float`, `double`, `BigDecimal` etc. This makes everything simpler, more consistent, and removes the need for unnecessary conversions and casting...

I'll still need to check for covarge holes in the new and changed code, and add unit tests to these to bring coverage back up to where it was. And I'll keep adding Javadoc also... But, I'm not planning any more major changes -- only bug fixes at this point. (Or if you convince me that something should not be done the way it is... :smile:)

Any comments, citicism or suggestions are welcome. You can add these into #170, if you will. Thanks in advance!

**Comment by rmathar** [comment]
Since roughly half a year slack is no longer supporting my operating system (openSUSE 15.5 currently) nor my browser (firefox 115.7 currently). So slack has become a no-go area for me.

**Reply by attipaci** [reply]
Yes, that's a bummer. Slack said they'd axe Fedora too (which is what I use), but I guess the did a U-turn on that after they realized it's actually useful for RHEL, which they allegedly want to support. Maybe either the OS support or the browser support will come back for you. In any case, there has not been much activity on the nom-tam-fits Slack workspace. Most of the discussions seem to be happening here on GitHub...

**Reply by olebole** [reply]
Debian testing currently has firefox 115.7.0esr and Slack works on it. However I would also agree that Slack isn't the best place for nom-tam-fits discussion.

**Comment by attipaci** [comment]
I will close the Slack workspace. I thought it might be useful to brainstorm about steering the direction of the library, but it looks like we area managing just fine without it.
